src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBinderSourceGenerator.Emitter.cs

```diff
@@ -0,0 +1,469 @@
+
+                foreach (KeyValuePair<ITypeSymbol, Location> pair in _targetTypes.TypesForGetMethodGen)
+                {
+                    ITypeSymbol type = pair.Key;
+                    string typeDisplayString = type.ToDisplayString();
+
+                    _writer.WriteBlockStart($"if (typeof(T) == typeof({typeDisplayString}))");
+
+                    if (UseStringBasedParse(type))
+                    {
+                        Emit_NotSupportedException_UnableToBindType(type, NotSupportedReadon_OnlyComplexObjectsSupported);
+                    }
+                    else if (TryEmitBindLogicFromIConfiguration(type, LiteralFor_obj, InitializationKind.Declaration))
+                    {
+                        _writer.WriteLine($"return (T)(object){LiteralFor_obj};");
+                    }
+
+                    _writer.WriteBlockEnd();
+                }
+
+                Emit_NotSupportedException_UnableToBindType(NotSupportedReason_TypeNotDetectedAsInput);
+                _writer.WriteBlockEnd();
+            }
+
+            private void EmitBindMethods()
+            {
+                foreach (KeyValuePair<ITypeSymbol, Location> pair in _targetTypes.TypesForBindMethodGen)
+                {
+                    EmitBindMethod("internal", pair.Key);
+                }
+
+                while (_privateBindMethodGen_QueuedTypes.Count > 0)
+                {
+                    EmitBindMethod("private", _privateBindMethodGen_QueuedTypes.Dequeue());
+                }
+            }
+
+            private void EmitBindMethod(string visibility, ITypeSymbol type)
+            {
+                if (_bindMethodGen_ProcessedTypes.Contains(type))
+                {
+                    return;
+                }
+
+                _bindMethodGen_ProcessedTypes.Add(type);
+
+                _writer.WriteBlockStart(@$"{visibility} static void {LiteralFor_Bind}(this {GlobalNameForIConfigurationType} {LiteralFor_configuration}, {type.ToDisplayString()} {LiteralFor_obj})");
+
+                if (type is IArrayTypeSymbol { } arrayType)
+                {
+                    EmitBindMethodImplForArray((arrayType.ElementType as INamedTypeSymbol)!);
+                }
+                else if (TypesAreEqual(type, _typeSymbolForIConfigurationSection))
+                {
+                    EmitCastToIConfigurationSection();
+                    EmitAssignment(LiteralFor_obj, LiteralFor_section);
+                }
+                else if (type is INamedTypeSymbol namedType)
+                {
+                    if (!type.IsValueType)
+                    {
+                        EmitCheckForNullArgument();
+                    }
+
+                    if (!IsCollection(namedType))
+                    {
+                        EmitBindMethodImplForObject(namedType);
+                    }
+                    else
+                    {
+                        CollectionBindingSpec? spec = GetCollectionBindingSpec(namedType);
```

Comment:
>GetCollectionBindingSpec [](http://example.com/codeflow?start=54&length=24)

Are we handling `ISet<T>` and `HashSet<T>` correctly?

Reply:
Yes.